sdk/src/main/java/org/coralibre/android/sdk/internal/device_info/DeviceList.java

```diff
@@ -0,0 +1,184 @@
+            this.oemDeviceFindings = Collections.unmodifiableList(oemDeviceFindings);
+            this.oemModelFindings = Collections.unmodifiableList(oemModelFindings);
+        }
+    }
+
+    public DeviceList(Context context, SharedPreferences sharedPrefs) throws IOException {
+        deviceList = loadDeviceInfoTable(context);
+        sharedPreferences = sharedPrefs;
+    }
+
+    public static DeviceInfo[] loadDeviceInfoTable(Context context) throws IOException {
+        BufferedReader infoCVS = new BufferedReader(new InputStreamReader(
+                context.getResources().openRawResource(LIST_FILE_ID)));
+        DeviceInfo[] deviceList = new DeviceInfo[LIST_LINE_COUNT - 1];
+        String line = infoCVS.readLine(); // discard the first line as it just contains the header
+        int i = 0;
+        while ((line = infoCVS.readLine()) != null) {
+            deviceList[i] = new DeviceInfo(line);
+            i++;
+        }
+        infoCVS.close();
+        return deviceList;
+    }
+
+    public DeviceInfo getInfo(final String manufacturer, final String device, final String model) {
+        int index = Arrays.binarySearch(deviceList, new DeviceInfo(manufacturer, device, model, 0, 0, NONE));
+        if (0 < index && deviceList[index].getModel().equals(model)) {
+            return deviceList[index];
+        } else {
+            return new DeviceInfo(manufacturer, device, model,
+                    DEFAULT_INFO.getRssiCorrection(),
+                    DEFAULT_INFO.getTx(),
+                    DEFAULT_INFO.getCalibrationConfidence());
+        }
+    }
+
+    private FindingsResult findingsInList(final String manufacturer, final String device, final String model) {
+        ArrayList<DeviceInfo> manufacturerFindings = new ArrayList<>();
+        ArrayList<DeviceInfo> oemDeviceFindings = new ArrayList<>();
+        ArrayList<DeviceInfo> oemModelFindings = new ArrayList<>();
+
+        for (DeviceInfo i : deviceList) {
+            if (manufacturer.toLowerCase().equals(i.getManufacturer().toLowerCase())) {
+                manufacturerFindings.add(i);
+
+                if (device.toLowerCase().equals(i.getDevice().toLowerCase())) {
+                    oemDeviceFindings.add(i);
+                }
+                if (model.toLowerCase().equals(i.getModel().toLowerCase())) {
+                    oemModelFindings.add(i);
+                }
+            }
+        }
+        return new FindingsResult(manufacturerFindings, oemDeviceFindings, oemModelFindings);
+    }
+
+    private static DeviceInfo getAverageOfFindings(List<DeviceInfo> findings) {
+        boolean allConfidenceGreaterLow = true;
+        int rssiSum = 0;
+        int txSum = 0;
+        for (DeviceInfo i : findings) {
+            rssiSum += i.getRssiCorrection();
+            txSum += i.getTx();
+            if (i.getCalibrationConfidence() == LOW) allConfidenceGreaterLow = false;
```

Comment:
Shouldn't this also check whether the confidence is `NONE`? Or is it impossible?

Reply:
Confidence `NONE` can't happen. It's the "fourth" level we introduced to show that something has no confidence at all. All you can find in the list is either `LOW`, `MEDIUM` or `HIGH`.